keps/808-GangConfig/README.md

```diff
@@ -0,0 +1,444 @@
+- Set the `minCount` in the `Gang` to `#pods per Job (parallelism)`.
+
+#### Pod Workload Reference Creation
+
+The `constructJobTemplate()` method should set the `workloadRef` field on the generated pod spec to reference the new Workload object. This field is how the `kube-scheduler` knows which pods to schedule as a gang.
+- If `jobSetSpec.GangConfig.GangMode == GangModeGang`,  all pods have a `workloadRef` object reference with the same pod group name and `PodGroupReplicaIndex` is unset.
+- If `replicatedJob.GangConfig.GangMode == GangModeGang`, all pods across all Job replicas for that ReplicatedJob have a `workloadRef` object reference with a different pod group name per ReplicatedJob and `PodGroupReplicaIndex` is unset.
+- If `replicatedJob.GangConfig.GangMode == GangModeReplicatedGang`, all pods for a single Job replica have a `workloadRef` object reference with a different pod group name per ReplicatedJob and `PodGroupReplicaIndex` is set to the Job replica index.
+
+#### Workload Deletion
+
+When the JobSet is deleted, the JobSet controller will also delete the Workload it created for that JobSet.
+
+#### Defaulting/Validation
+
+- `jobSetSpec.GangConfig.GangMode` and `replicatedJob.GangConfig.GangMode` cannot be specified at once (except to `GangModeOff`).
+- `jobSetSpec.GangConfig.GangMode` and `replicatedJob.GangConfig.GangMode` are immutable.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+#### Unit Tests
+controllers: 01/19/2024 - 30.2%
+Unit tests are low in this area as we do more of our testing at the integration level.
+
+`jobset_controller.go` and `jobset_controller_test.go` will be extended to handle the new `GangConfig` fields, set the `workloadRef` in Pod specs & generate a Workload object.
+
+#### Integration tests
+1. Gang scheduling for a single or multiple ReplicatedJobs with JobSet controller created Workload objects.
+2. Gang scheduling for a single or multiple Job replicas with JobSet controller created Workload objects.
+3. Gang scheduling for the entire JobSet with JobSet controller created Workload objects (when this API is supported).
+4. Gang scheduling at JobSet, ReplicatedJob and Job replica levels when the JobSet is being recreated.
+5. Gang scheduling for ReplicatedJobs when DependsOn or StartupPolicy is configured.
+
+In all tests, validate Workload spec values for `controllerRef` and `podGroups` and JobSet Pod spec values for `workloadRef`. 
+
+Additionally, manually validate the cases for user created Workload objects.
+
+### Graduation Criteria
+
+- `kube-scheduler` changes and `Workload` API alpha are targeted to `1.35`.
+- Support for `replicatedJob.GangConfig` will be alpha in 1.35 behind an alpha feature gate and graduated to stable when `Workload` API is stable.
```

Comment:
We should have a feature gate for JobSet with this KEP that matches the name of the gang scheduling feature gate.

But we will want to keep this feature disabled by default until workload API is stable on all supported k8s versions of JobSet.

We will need to introduce feature gate handling in JobSet but I think this feature warrants this if we want to merge this in 1.35.